### PR TITLE
MCOL-5429 Fix high memory consumption in GROUP_CONCAT() processing.

### PR DESCRIPTION
**Pull request by tntnatbry**
1. Input and output RowGroup's used in GROUP_CONCAT classes are currently allocating a raw memory buffer of size equal to the actual width of the string datatype. As an example, for the following query:
  SELECT col1, GROUP_CONCAT(col2) FROM t GROUP BY col1;
If col2 is a TEXT field with default width, the input RowGroup containing the target rows to be concatenated will assign 64kb of memory for every input row in the RowGroup. This is wasteful as actual field values in real workloads would be much smaller. We fix this by enabling the RowGroup to use the StringStore when the RowGroup contains long strings.

2. RowAggregation::initialize() allocates a memory buffer for a NULL row. The size of this buffer is equal to the row size for the output RowGroup. For the above scenario, using the default group_concat_max_len (which is a server variable that sets the maximum length of the GROUP_CONCAT string) value of 1mb, the buffer size would be
(1mb + 64kb + some additional metadata). If the user sets group_concat_max_len to a higher value, say 3gb, this buffer size would be ~3gb. Now if the runtime initiates several instances of RowAggregation, total memory consumption by PrimProc could exceed the hardware memory limits causing the OS OOM to kill the process. We fix this problem by again enabling the StringStore for the NULL row allocation.

3. In the plugin code in buildAggregateColumn(), there is an integer overflow when the server group_concat_max_len variable (which is an uint32_t) is set to a value > INT32_MAX (such as 3gb) and is assigned to
CalpontSystemCatalog::ColType::colWidth (which is an int32_t). As a short term fix, we saturate the assigned value to colWidth to INT32_MAX. Proper fix would be to upgrade
CalpontSystemCatalog::ColType::colWidth to an uint32_t.

### CODE REVIEW

**Comment by tntnatbry** [comment]
Please don't merge until QA tests the crashing cases on this PR build.

**Review by mariadb-SergeyZefirov — APPROVED** [review]
I agree with the changes.

File: dbcon/mysql/ha_mcs_execplan.cpp
Comment (drrtuy) [inline dbcon/mysql/ha_mcs_execplan.cpp:5361]: Does INT32_MAX const come from MDB namespaces?

File: utils/rowgroup/rowaggregation.cpp
Comment (drrtuy) [inline utils/rowgroup/rowaggregation.cpp:694]: I didn't get it. Do you add fNullRowGroup to just use it for init here? Why then do you need to have the fNullRowGroup as RowAggregation attribute?

**Comment by drrtuy** [comment]
The patch LGTM. There are some questions to be clarified before we merge the patch though.
A separate but relevant observation. There is clone of GroupConcat to deliver JSON_ARRAYAGG functionality. Could you @tntnatbry  plz look if it is also affected but the inefficient memory allocations so we can file another Jira ?

File: dbcon/mysql/ha_mcs_execplan.cpp
Comment (tntnatbry) [reply dbcon/mysql/ha_mcs_execplan.cpp:5361]: No, this is the c const literal defined in the `<cstdint>` header.

File: utils/rowgroup/rowaggregation.cpp
Comment (tntnatbry) [reply utils/rowgroup/rowaggregation.cpp:694]: Yes, fNullRowGroup is used here to init a single row RGData with StringStore enabled, and then make Row point to this raw data via fNullRowGroup.

**Comment by tntnatbry** [comment]
@drrtuy Ok, let me look into the JSON_ARRAYAGG scenario.

File: dbcon/mysql/ha_mcs_execplan.cpp
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_execplan.cpp:5361]: GTK.
JFYI Modern cpp favors std::numeric_limits<int32_t>::max() instead of those constants.

File: utils/rowgroup/rowaggregation.cpp
Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:694]: IMHO It is worth to limit the span of this RG and remove it from the list of RowAggregation attributes.

File: dbcon/mysql/ha_mcs_execplan.cpp
Comment (tntnatbry) [reply dbcon/mysql/ha_mcs_execplan.cpp:5361]: That's correct. Thanks for highlighting the use of `std::numeric_limits` here.

File: utils/rowgroup/rowaggregation.cpp
Comment (tntnatbry) [reply utils/rowgroup/rowaggregation.cpp:694]: `fNullRowData` which is also just used in the else block is also an attribute of RowAggregation. We can keep this as is to maintain the code semantics here.

Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:694]: Feel free to remove fNullRowData also b/c these two are wastes RAM in runtime.

Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:694]: Or you can just save some some bytes not following the wrecked pattern and left fNullRowData intact if you don't want to increase the scope.

Comment (tntnatbry) [reply utils/rowgroup/rowaggregation.cpp:694]: No. This is not wrecked pattern. The reason fNullRowData is an attribute of the class is because it owns a memory buffer to which fNullRow points. If we remove fNullRowData as a class member and make it local variable with the scope of the else block, then the RGData dtor would get called and destroy the underlying memory buffer, causing fNullRow to point to an invalid memory address.

Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:694]: Makes sense. You don't need RG to stay around as RowStorage attribute though, right?

Comment (tntnatbry) [reply utils/rowgroup/rowaggregation.cpp:694]: `fRowGroup` is tightly coupled with `fRowRGData`, so I am keeping the declarations of both at the same place.

Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:694]: But this tight coupling spends 312 bytes per each RowAggregation instance just for nothing and you use RG only once!

Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:694]: I didn't get your logic behind this. IMHO this approach values the form more then the semantics but I won't push on this anymore.

**Review by drrtuy — APPROVED** [review]

File: utils/rowgroup/rowaggregation.cpp
Comment (tntnatbry) [reply utils/rowgroup/rowaggregation.cpp:694]: If I make `fNullRowGroup` local to the scope of the if block, now `fNullRowGroup.initRow(&fNullRow);` would be invalid. This is because `initRow()` is making some `Row` pointer attributes point to the `RowGroup` objects (`Row::oldOffsets`, `Row::stOffsets` are pointers to their corresponding objects defined in the `RowGroup` class). As soon as `fNullRowGroup` goes out of scope, these `offset` objects are destroyed and the corresponding `Row` pointers to these offsets would now be invalid.

Below is the `initRow` definition. If you look at the if-else blocks, you will see the pointer assignments to the offset attributes:
```c++
void RowGroup::initRow(Row* r, bool forceInlineData) const                                                                                                                                 
{                                                                                                                                                                  
  r->columnCount = columnCount;                                                                                                                                    
                                                                                                                                                                                   
  if (LIKELY(!types.empty()))                                                                                                                                                                                                                       
  {                                                                                                                                                                                                                                      
    r->colWidths = (uint32_t*)&colWidths[0];                                                                                                                                                                                       
    r->types = (execplan::CalpontSystemCatalog::ColDataType*)&(types[0]);                                                                                                
    r->charsetNumbers = (uint32_t*)&(charsetNumbers[0]);                                                                                                                                                                                            
    r->charsets = (CHARSET_INFO**)&(charsets[0]);                                                                                                                         
    r->scale = (uint32_t*)&(scale[0]);                                                                                                                                                                                                              
    r->precision = (uint32_t*)&(precision[0]);                                                                                                                      
  }                                                                                                                                                                
                                                                                                                                                                                                     
  if (forceInlineData)                                                                                                                                              
  {                                                                                                                                                                                                                                                 
    r->useStringTable = false;                                                                                                                                                                                       
    r->oldOffsets = (uint32_t*)&(oldOffsets[0]);                                                                                                                   
    r->stOffsets = (uint32_t*)&(stOffsets[0]);                                                                                                                                                
    r->offsets = (uint32_t*)&(oldOffsets[0]);                                                                                                                                                           
  }                                                                                                                                                                
  else                                                                                                                                                                                     
  {                                                                                                                                                                
    r->useStringTable = useStringTable;                                                                                                                                            
    r->oldOffsets = (uint32_t*)&(oldOffsets[0]);                                                                                                                                                                                                    
    r->stOffsets = (uint32_t*)&(stOffsets[0]);                                                                                                                                                                                           
    r->offsets = offsets;                                                                                                                                                                                                          
  }                                                                                                                                                                      
                                                                                                                                                                                                                   
  r->hasLongStringField = hasLongStringField;                                                                                                                             
  r->sTableThreshold = sTableThreshold;                                                                                                                             
  r->forceInline = forceInline;                                                                                                                                    
  r->hasCollation = hasCollation;                                                                                                                                                                                            
}
```

Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:694]: This RowGroup placement makes sense to me now. Thank you for the patience and the detailed explanation.

**Pull request merged**
